drivers/nec/display.cr

```diff
@@ -0,0 +1,308 @@
+      else
+        return task.try &.abort("-- NEC LCD, command failed: #{command}\n-- NEC LCD, response was: #{ascii_string}")
+      end
+    end
+
+    task.try &.success
+  end
+
+  private def parse_response(data : String)
+    # 14..15 == type (we don't care)
+    value = data[20..23].to_i(16)
+    command = Command.from_value(data[10..13].to_i(16))
+
+    logger.debug { "command is 0x#{data[10..13]}" }
+    logger.debug { command }
+    logger.debug { "value is 0x#{data[20..23]}" }
+    logger.debug { value }
+
+    case command
+    when .video_input?
+      self[:input] = Input.from_value(value)
+    when .audio_input?
+      self[:audio] = Audio.from_value(value)
+    when .volume_status?
+      self[:volume] = value
+      self[:audio_mute] = value == 0
+    when .brightness_status?
+      self[:brightness] = value
+    when .contrast_status?
+      self[:contrast] = value
+    when .mute_status?
+      self[:audio_mute] = value == 1
+      self[:volume] = 0 if value == 1
+    when .auto_setup?
+      # auto_setup
+      # nothing needed to do here (we are delaying the next command by 4 seconds)
+    else
+      logger.info { "-- NEC LCD, unknown response: #{data[10..13]}" }
+      logger.info { "-- NEC LCD, full response was: #{data}" }
+    end
+  end
+
+  # Types of messages sent to and from the LCD
+  enum MsgType : UInt8
+    Command           = 0x41 # 'A'
+    CommandReply      = 0x42 # 'B'
+    GetParameter      = 0x43 # 'C'
+    GetParameterReply = 0x44 # 'D'
+    SetParameter      = 0x45 # 'E'
+    SetParameterReply = 0x46 # 'F'
+  end
+
+  enum Command
+    VideoInput       = 0x0060
+    AudioInput       = 0x022E
+    VolumeStatus     = 0x0062
+    MuteStatus       = 0x008D
+    PowerOnDelay     = 0x02D8
+    ContrastStatus   = 0x0012
+    BrightnessStatus = 0x0010
+    AutoSetup        = 0x001E
+    PowerQuery       = 0x01D6
+    Save             = 0x0C
+  end
+
+  private def format_value(value : Int, length : Int = 4) : String
+    length = 2 if value == 0x0C # To deal with Command::Save
+    value.to_s(16, true).rjust(length, '0')
+  end
+
+  {% for name in Command.constants %}
+    @[Security(Level::Administrator)]
+    def {{name.id.underscore}}(priority : Int32 = 0)
+      do_send(MsgType::GetParameter, Command::{{name.id}}, priority: priority, name: {{name.id.underscore.stringify}})
+    end
+  {% end %}
+
+  private def check_checksum(data : Bytes)
+    checksum = 0x00_u8
+    # Loop through the second to the third last element
+    if data.size >= 2
+      data[1..-3].each do |b|
+        checksum = checksum ^ b
+      end
+      # Check the checksum equals the second last element
+      logger.debug { "Error: checksum should be #{checksum.to_s(16)}" } unless checksum == data[-2]
+      checksum == data[-2]
+    else
+      true
+    end
+  end
+
+  # Builds the command and creates the checksum
+  private def do_send(type : MsgType, data : Command | Array(Int), **options)
```

Comment:
There's a couple of things going on in here where it may be good to take a slightly different approach and neaten up.

The passing of the message type and an accompanying payload is nice and looks to match the structure of the device protocol well. Let look at the payload itself though - the only common abstraction here between _all_ message types is it is just a list of bytes. With each message type though there is a common structure so rather than handling everything here, this method could take a `MsgType` and a `data : Bytes`, then purely focus on wrapping these into the full command packet (forming the heading, adding the framing to the message, checksumming etc).

Upstream on this you could then have three methods `set_parameter`, `get_parameter` and `exec_command` which take args as appropriate and purely focus on forming up the `data` that's passed here.

Reply:
If you go this route, it should help neaten up some of the magic nums that are still in use for the power control.